solr/core/src/java/org/apache/solr/search/CaffeineCache.java

```diff
@@ -114,9 +126,12 @@ public Object init(Map<String, String> args, Object persistence, CacheRegenerato
     str = args.get(MAX_RAM_MB_PARAM);
     int maxRamMB = str == null ? -1 : Double.valueOf(str).intValue();
     maxRamBytes = maxRamMB < 0 ? Long.MAX_VALUE : maxRamMB * 1024L * 1024L;
-    str = args.get(CLEANUP_THREAD_PARAM);
-    cleanupThread = str != null && Boolean.parseBoolean(str);
-    if (cleanupThread) {
+    cleanupThread = Boolean.parseBoolean(args.get(CLEANUP_THREAD_PARAM));
+    async = Boolean.parseBoolean(args.get(ASYNC_PARAM));
+    if (async) {
+      // We record futures in the map to decrease bucket-lock contention, but need computation handled in same thread
+      executor = Runnable::run;
+    } else if (cleanupThread) {
       executor = ForkJoinPool.commonPool();
```

Comment:
You kind of narrowly dodge this being part of this issue, so more of an FYI, I'll second Ben's comments on this common thread pool usage.

It's a trappy and confusing use of common pool indicated to only impact a "cleanup" thread.

This is in fact a general executor for Caffeine, not simply the cleanup thread pool. In most cases, that is not a big deal, but the common pool is extremely easy to put into thread starvation, especially the common pool, if you perform any blocking tasks on it. There is an ugly workaround where you can indicate to the pool that you are going to block so it can compensate while you do so, but in general this pool is entirely meant for non blocking async type tasks. Where it get's trappy is that it's also used to do the work of loading for the async cache, which will often be doing IO, which is blocking. In this issue, async will use the current thread, so the case is avoid, but the code remains unclear and trappy and Caffeine can shift on where that executor is used in future versions given it's not a "cleanup" executor.

Reply:
Yea, I detour around it, and this issue existed previously. We might want to investigate it in a follow up issue, but I'd rather not get into it now.